include/toolchain.h

```diff
@@ -41,4 +41,18 @@
 #include <toolchain/other.h>
 #endif
 
+/*
+ * Ensure that __BYTE_ORDER__ and related preprocessor definitions are defined,
+ * as these are often used without checking for definition and doing so can
+ * cause unexpected behaviours.
+ */
+#ifndef _LINKER
+#if !defined(__BYTE_ORDER__) || !defined(__ORDER_BIG_ENDIAN__) || \
```

Comment:
Should we have here a check that makes sure that `__BYTE_ORDER__ `is set and either `__ORDER_BIG_ENDIAN__` or `__ORDER_LITTLE_ENDIAN__`. So something like this:
```
#if !(defined(__BYTE_ORDER__) && (defined(__ORDER_BIG_ENDIAN__) || \
    defined(__ORDER_LITTLE_ENDIAN__)))
```

Reply:
The preprocessor definitions __ORDER_BIG_ENDIAN__ and __ORDER_LITTLE_ENDIAN__ must be defined regardless of whether the system is big-endian or little-endian.

This is because these two definitions are constants which will be compared against __BYTE_ORDER__, rather than stating whether the system's endianness.

See below:
```
$  gcc -dM -E - < /dev/null
#define __ORDER_LITTLE_ENDIAN__ 1234
#define __ORDER_BIG_ENDIAN__ 4321
#define __BYTE_ORDER__ __ORDER_LITTLE_ENDIAN__
...
```

Reply:
All 3 should be defined, not just 2. Otherwise, we may implicitly fall into an #else case. This PR as it stands also requires all 3 to be defined, which is what we want.

Reply:
> All 3 should be defined, not just 2. Otherwise, we may implicitly fall into an #else case. This PR as it stands also requires all 3 to be defined, which is what we want.

Ok, makes sense.